graphsync/graphsync.md

```diff
@@ -0,0 +1,151 @@
+- `dag` - a directed acyclic graph. For our purposes, our DAGs are all IPLD (connected by hash links, authenticated, content addressed, etc.)
+- `selector` - an expression that identifies a specific subset of a graph.  ($$ S(G) \subset G $$)
+- `selector language` - the language defining a family of selectors
+- `request` - a request for content from one `peer` to another. This is similar to HTTP, RPC, or API requests.
+- `response` - the content sent from `responder` to `requester` fulfilling a `request`.
+- `requester` - the peer which initiates a `request` (wants content).
+- `responder` - the peer receiving a `request`, and providing content in a `response` (provides content).
+- `request process` - a request and its fulfillment is a sub-process, a procedure call across peers with the following phases (at a high level):
+  - (1) The `requester` initiates by sending a request message (`req`) to the `responder`, specifying desired content and other request parameters.
+  - (2) Upon receiving a request message, the `responder` adds the request to a set of active requests, and starts processing it.
+  - (3) The `responder` fulfills the request by sending content to the `requester` (the `response`) .
+  - (4) The `responder` and `requester` can terminate the request process at any time.
+  - Notes:
+    - We are explicitly avoiding the `client-server` terminology to make it clear that `requester` and `responder` are "roles" that any peer might play, and to avoid failling in the two-sided client-server model of the web..
+    - `requests` may be short or long-lived -- requests may be as short as microseconds or last indefinitely.
+- `priority` - a numeric label associated with a `request` implying the relative ordering of importance for requests. This is a `requester's` way of expressing to a `responder` the order in which the `requester` wishes the `requests` to be fulfilled. The `responder` SHOULD respect `priority`, though may return `responses` in any order.
+
+## Interfaces
+
+This is a listing of the data structures and process interfaces involved in the graphsync protocol. For simplicity, we use Go type notation, though of course graphsync is language agnostic.
+
+```go
+type Graphsync interface {
+	Request(req	Request) (Response, error)
+}
+
+type Request struct {
+    Selector Selector
+    Priority Priority      // optional
+    Expires  time.Duration // optional
+}
+
+type GraphSyncNet interface {
+    SendMessage(m Message)
+    RecvMessage(m Message)
+}
+```
+
+
+## Network Messages
+
+```protobuf
+message GraphsyncMessage {
+
+  message Request {
+    int32 id = 1;       // unique id set on the requester side
+    bytes selector = 2; // ipld selector to retrieve
+    bytes extra = 3;    // aux information. useful for other protocols
+    int32 priority = 4;	// the priority (normalized). default to 1
+    bool  cancel = 5;   // whether this cancels a request
+  }
+
+  message RequestList {
```

Comment:
We can probably remove this indirection and embed the requests direly in the message.